### ISSUE TITLE
Datastore feature request: batch operation of mixed types

### ISSUE THREAD

**Issue by timanovsky** [GoogleCloudPlatform/gcloud-ruby#646]
Something like this:
```ruby
      def update_delete(items_to_update, items_to_delete = nil)
        mutation = Proto.new_mutation
        save_entities_to_mutation items_to_update, mutation

        if items_to_delete.present?
          keys_to_delete = items_to_delete.map do |e_or_k|
            e_or_k.respond_to?(:key) ? e_or_k.key.to_proto : e_or_k.to_proto
          end
          mutation.delete = keys_to_delete
        end

        response = connection.commit mutation
        auto_id_assign_ids response.mutation_result.insert_auto_id_key
        # Make sure all entity keys are frozen so all show as persisted
        items_to_update.each { |e| e.key.freeze unless e.persisted? }
        items_to_update
      end
```

**Comment by blowmage** [comment]
@timanovsky Thanks for opening this issue. Don't we already have this functionality with `Dataset#transaction`?

```ruby
dataset.transaction do |tx|
  tx.save *items_to_update
  tx.delete *items_to_delete
end
```

What is the use-case for doing this outside of a transaction?

**Comment by timanovsky** [comment]
Sometimes you don't need txn guaranties, and this is way way faster than
transaction in my tests. 3-4 items transaction can take up to 700ms, as a
batch operation it takes may be 200 ms.

Also batch operations allows more flexible set of unrelated items (I think
txn is limited to 25 different ancestors).

Best regards,
Alexey Timanovsky.

**Comment by blowmage** [comment]
@quartzmo, what do you think about adding a batch-mode update outside of `transaction`? We do something similar for [`DNS Zone#update`](https://googlecloudplatform.github.io/gcloud-ruby/#/docs/v0.8.0/gcloud/dns/zone?method=update-instance).

I don't know that I'd name the Datastore method `update`, since we already have `save` and `delete`, and it might lead folks to think this is a CRUD update. Something like `batch`?

```ruby
dataset.batch items_to_update, items_to_delete

# or

dataset.batch do |b|
  b.save items_to_update
  b.delete items_to_delete
end
```

**Comment by quartzmo** [comment]
Here is my *opinion*: In the service API (Rest and now gRPC), support for a "non-transactional" [commit](https://cloud.google.com/datastore/reference/rpc/google.datastore.v1beta3#google.datastore.v1beta3.CommitRequest) has been tacked on to the apparatus for transactions. It's placement is not elegant, but it is a useful feature. Now, I know the example below would open us up for some ridicule, just as we can criticize the service API, but why should we not just follow their lead? This approach adds the feature without adding a new method to our API, which I think is worth the damage to conceptual integrity. Maybe we could alias `batch` to `transaction`?

```ruby
dataset.transaction mode: :non_transactional do |tx|
  tx.save *items_to_update
  tx.delete *items_to_delete
end
```

same as:


```ruby
dataset.batch mode: :non_transactional do |tx|
  tx.save *items_to_update
  tx.delete *items_to_delete
end
```

**Comment by blowmage** [comment]
Yeah, that's definitely something new. In that case I think a non-transactional commit would/should look like this:

```ruby
dataset.commit do |c|
  c.save *items_to_update
  c.delete *items_to_delete
end
```

We could even support this on `Transaction#commit` as well.

```ruby
tx = dataset.transaction
begin
  tx.commit do |c|
    c.save *items_to_update
    c.delete *items_to_delete
  end
rescue
  tx.rollback
end
```

Thoughts?

**Comment by quartzmo** [comment]
My opinion is, we should avoid adding new methods to our API. Which is why i think we should just follow their lead and add a flag where it doesn’t fit conceptually. I see it as a power-user feature, kinda buried in the docs, not bothering the new user much at all. Adding a new method imposes more burden on the new user, imo.

**Comment by blowmage** [comment]
I think its okay to add a new method if we are adding new functionality. Commit is a central part of the API in v1beta2 and v1beta3, and I think it makes more sense for users to type `ds.commit do` than `ds.transaction :non_transactional do`.

**Comment by blowmage** [comment]
Especially since we'd need to pass a different object than Transaction to the block since Transaction always calls `begin_transaction` to get a transaction ID. I think the behavior is different enough that we should separate the implementations.

**Comment by quartzmo** [comment]
@blowmage I see your point about the coupling of `Transaction` to its `id`. Sounds good to me.

**Comment by blowmage** [comment]
@timanovsky This has also been released in [0.8.1](https://rubygems.org/gems/gcloud/versions/0.8.1). Thanks again!

**Comment by timanovsky** [comment]
Will try today, this is great addition together with BLOBs support!

Best regards,
Alexey Timanovsky.

**Comment by timanovsky** [comment]
Hi,

I have a problem updating to  0.8.1 gem:

```
> bundle update gcoud

....

Could not find google-protobuf-3.0.0.alpha.5.0.5.1 in any of the sources

```

Best regards,
Alexey Timanovsky.

**Comment by timanovsky** [comment]
Update: This is because of upstream problem: https://github.com/grpc/grpc/issues/6427

**Comment by blowmage** [comment]
Can you re-file that issue against google/protobuf?

**Comment by timanovsky** [comment]
@blowmage I have not found protobuf on github

**Comment by blowmage** [comment]
@timanovsky https://github.com/google/protobuf

**Comment by timanovsky** [comment]
Thanks, created: https://github.com/google/protobuf/issues/1485

Best regards,
Alexey Timanovsky.

**Comment by blowmage** [comment]
Thanks! They are on it. There was a problem with the Darwin gem, and so they yanked it late last night and updated that one only.

**Comment by quartzmo** [comment]
@timanovsky We just released [0.8.2](https://rubygems.org/gems/gcloud/versions/0.8.2) to fix a bug in 0.8.1 (see #667 for details.) Can you please update to 0.8.2 as soon as possible? Thank you.

**Comment by timanovsky** [comment]
Thanks for letting know!

Best regards,
Alexey Timanovsky.
